File: codes/programmers/탑.cpp
#include <string>
#include <vector>
#include <stack>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> heights) {
    vector<int> answer;
    stack<int> st;
    bool check = true;
    for(int i = heights.size()-1; i >= 0; i--)
    {
        for(int j = i - 1; j >= 0; j--)
        {
            if(heights.at(i) < heights.at(j))
            {
                st.push(j+1);
                check = false;
                break;
            }
        }
        if(check)
            st.push(0);
        check = true;
    }
    while(!st.empty())
    {
        answer.push_back(st.top());
        st.pop();
    }
    
    return answer;
}

void print(vector<int> heights)
{
    vector<int> t = solution(heights);
    for(auto &e : t)
        cout << e << " ";
    cout << endl;

}

int main()
{
    print({6,9,5,7,4});
    print({3,9,9,3,5,7,2});
    print({1,5,3,6,7,6,5});
}

File: PS_Study/week_02/yoon/Programmers/탑.py
def solution(heights):
    answer = [0] * len(heights)
    
    for i in range(len(heights)-1, 0, -1):
        for j in range(i-1, -1, -1):
            if heights[i] < heights[j]:
                answer[i] = j+1
                break
    return answer

"""
1. 높이가 같은 탑은 수신 안함
2. Stack 인 것 같은데 stack을 쓴다면?

O(N^2) 보다 복잡도가 낮은 풀이가 있나?

"""

print(solution([6,9,5,7,4]))